### ISSUE TITLE
Moved AD Users result in new Rancher Users

### ISSUE THREAD

**Issue by samjustus** [rancher/rancher#37994]
SURE-4716

**Rancher Server Setup**
- Rancher version: 2.5.4

**Describe the bug**
If Rancher is configured to use AD for authentication, and AD user accounts get moved between OUs, then when the users log back into Rancher a new user is created and all their old permissions are not preserved.

**To Reproduce**

1. Enable Rancher authentication to work with AD
2. Create a new user
3. Give user access to various projects
4. Move user to another OU in AD
5. Login as same user

**Result**
New user is created and permissions are not preserved.

**Expected Result**
New user should not be created so that cleanup work in workaround shouldn't be necessary.

**What is the workaround:**

Note the 'principalIds' value for the new user
Delete the new user from the Rancher UI (so we won't be left with duplicate 'principalIds'
Use kubectl commands to edit the user yaml for the original user and then change their 'principalIds' value to that of the new user that was created

**Comment by samjustus** [comment]
#21042
#17445
#29996
#13434

**Comment by mayrstefan** [comment]
One of the referenced issues is four years old and one of the old pull requests was rejected. Is there already a plan how this could be implemented in a better way?

**Comment by MbolotSuse** [comment]
Summary of the various conversations that have taken place on this issue (as well as what needs to be done):

Quick look at this (and the [ms docs](https://docs.microsoft.com/en-us/windows/win32/ad/naming-properties#objectsid)) says that we have two options for how to proceed here:
- Use the object GUID (global for each user/group)
- Use the object SID (can change if a user is moved across domains, but we can track the history of the users SIDs)

Either of these options would allow us to track a user in a more reliable fashion. The main issue here will be backwards compatibility with older versions of rancher which use the DN for users. I think that to accomplish this we will need to give users a user-triggered one time upgrade action, which converts the current principal ids into the new state.

Some notes on this:
- This upgrade will likely be one-time and not reversible (outside of a backup/restore operation)
- Documentation will need to be created so that we can instruct customers on the usage/impact of the tool
- Upgrade script will need to not only change the principals on a given user, but will also need to adjust the various CRTBs/PRTBs/GRBs and RBAC primitives (rolebindings/clusterrolebindings) to convert from the old state to the new state
- If we use the object SID (which seems to be the preferable option) we will need to track the SIDs for a user and occasionally update the principal on the user/the related bindings (as described above) to ensure that we are tracking the latest SID for a user. This will probably need a log statement/compliance trial so that admins are aware of why these values have changed.

**Comment by cbron** [comment]
We should check to see if this is related: https://github.com/rancher/rancher/issues/37403

**Comment by flrichar** [comment]
Similar in concern to [#29996](https://github.com/rancher/rancher/issues/29996) if the Group changes because an OU is removed, the principal mappings break.  It would be great to have a way to search/replace these mappings where the Group is essentially the same if the schema changes slightly.

**Comment by AndrewHoffmanQA** [comment]
Chad and I got together and tested the scenario described above and did some adhoc testing in general around changing user's  OU's. 

additionally we discussed testing the scenario where the users are on an older rancher image, upgrading the image and confirming the user's principleIDs are updated

**Comment by AndrewHoffmanQA** [comment]
completed the upgrade scenario testing as well, confirmed that when upgrading from 2.7.3 to 2.7.5-rc5 the principleIds updated as expected. 

all testing for this ticket has passed, I am moving it into 'done' column

**Comment by JacieChao** [comment]
I found a new issue with this change. Users can't add to cluster member/project member if the AD user is not under the User Search Base DN.

Env: Helm install on K3s for Rancher v2.7-head

## Steps to Reproduce

- Enable Rancher authentication with AD. User search base set to `OU=test,DC=jacie,DC=work`
- Give user qa(the user is based in `OU=test,DC=jacie,DC=work`) cluster member access to a test cluster. the qa user principal can be selected from the search input and also can save the permission.
- Give user dev(the user is based in `OU=sy,OU=test,DC=jacie,DC=work`) cluster member access to a test cluster. 
- The dev user principal can be found and selected from the search input, but the UI will get **`No identities can be retrieved`** error when clicking the save button.

## Expected Result

All users under the Search Base can be granted cluster/project permissions.

## More Information

I think the [search scope here](https://github.com/rancher/rancher/blob/release/v2.7/pkg/auth/providers/activedirectory/activedirectory_client.go#L332C1-L332C1) should be `ScopeWholeSubtree` for this situation because we use the [SearchBase](https://github.com/rancher/rancher/blob/release/v2.7/pkg/auth/providers/activedirectory/activedirectory_client.go#L372) to generate the search request. If using `ScopeSingleLevel` it can't get other users whose OU is not just under the SearchBase DN.

**Comment by prachidamle** [comment]
This issue was fixed as part of v2.7.5, but the data migration involved as part of the resolution has broken a lot of setups post v2.7.5 upgrade. 

In order to unblock users to upgrade to newer Rancher versions with the least risk, it has been decided to revert the above GUID fix that were added in v2.7.5
This ticket will be put in review for resolution in later release, we will be re-looking at fixing this issue back in a minor release instead of a patch release.

**Comment by MKlimuszka** [comment]
This is tied to the issue that was fixed in 2.7.5, which was then reverted. So this needs to be fixed again when we support searching in AD with GUID.

**Comment by samjustus** [comment]
sure-6552

**Comment by samjustus** [comment]
this ticket is no longer necessary with #44916

**Comment by enrichman** [comment]
Issue https://github.com/rancher/rancher/issues/44916 was closed because this bug will be fixed using a different approach than the one originally planned.

**Comment by Priyashetty17** [comment]
Das to pick this up for QA

**Comment by samjustus** [comment]
unfortunately fixing the issue migrating the principalID to something more robust is very problematic. We decided to have again a look at everything from the beginning, but nothing is planned yet.
